File: algorithms_on_graphs/toposort.java
import java.util.*;

public class toposort {
    private static ArrayList<Integer> toposort(ArrayList<Integer>[] adj) {
        boolean[] used = new boolean[adj.length];
        ArrayList<Integer> order = new ArrayList<>();
        for(int i = 0; i < adj.length; i++) {
            Deque<Integer> stack = new ArrayDeque<>();
            stack.push(i);
            dfs(adj, stack, used, order);
        }
        Collections.reverse(order);
        return order;
    }
    
    private static void dfs(ArrayList<Integer>[] adj,
                            Deque<Integer> stack,
                            boolean[] used,
                            ArrayList<Integer> order) {
        Set<Integer> visited = new HashSet<>();
        while(!stack.isEmpty()) {
            int current = stack.peek();
            if(used[current]) {
                stack.pop();
                continue;
            }
            if(visited.contains(current)) {
                order.add(current);
                stack.pop();
                used[current] = true;
            } else {
                if(!adj[current].isEmpty()) {
                    visited.add(current);
                    adj[current].forEach(edge -> {
                        if(!used[edge]) stack.push(edge);
                    });
                } else {
                    order.add(current);
                    stack.pop();
                    used[current] = true;
                }
            }
        }
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
        }
        List<Integer> order = toposort(adj);
        for (int x : order) {
            System.out.print((x + 1) + " ");
        }
    }
}

File: algorithms_on_graphs/toposort.js
// Input: First line "n m", where n is number of vertices and m number of edges. After this,
// m lines with "u v", where both are connected vertices of the graph.
// Example input: 4 3
//                1 2
//                4 1
//                3 1
// Output: Any topological ordering of the vertices
// Example output: 4 3 1 2

const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  terminal: false
});

const readLines = () => {
  process.stdin.setEncoding('utf8');

  const connections = [];

  rl.once('line', line => {
    const [vertices, edges] = line.toString().split(' ').map(v => parseInt(v, 10));
    let n = edges;

    if (edges === 0) {
      process.stdout.write(checkConnectedComponents(vertices, connections).toString());
      process.exit();
    }

    const readConnection = line => {
      const link = line.toString().split(' ').map(v => parseInt(v, 10));

      connections.push(link);

      if (!--n) {
        process.stdout.write(toposort(vertices, connections).toString());
        process.exit();
      }
    };

    rl.on('line', readConnection);
  });
};

function buildGraph(verticesQt) {
  const graph = {};

  for (let i = 1; i <= verticesQt; i++) {
    graph[i] = {
      edges: new Set(),
    };
  }

  return graph;
}

function createConnections(graph, connections) {
  connections.forEach(([origin, destiny]) => {
    graph[origin].edges.add(destiny);
  });
}

function dfs(graph, stack, used, order) {
  const visited = new Set();

  while(stack.length) {
    const current = stack[stack.length - 1];

    if (used.has(current)) {
      stack.pop();
      continue;
    }

    if (visited.has(current)) {
      order.push(current);

      stack.pop();

      used.add(current);

      continue;
    }

    if (graph[current].edges.size) {
        visited.add(current);

        graph[current].edges.forEach(edge => {
          if (!used.has(edge)) {
            stack.push(edge);
          }
        });
    } else {
      order.push(current);

      stack.pop();

      used.add(current);
    }
  }
}

function toposort(verticesQt, connections) {
  const graph = buildGraph(verticesQt);

  createConnections(graph, connections);

  const order = [];
  const used = new Set();

  for(let i = 1; i <= verticesQt; i++) {
    const stack = [];

    stack.push(i);

    dfs(graph, stack, used, order);
  }

  return order.reverse().join(' ');
}

function test(outputType, onlyTest) {
  let testCases = [
    {
      id: 1,
      run: () => toposort(
        6,
        [
          [1, 2],
          [2, 6],
          [3, 1],
          [3, 5],
          [4, 6],
          [5, 4],
        ]
      ),
      expected: '1 3 5 2 4 6'
    },
  ];

  if (onlyTest !== undefined) {
    testCases = [testCases.find(testCase => testCase.id == onlyTest)].filter(test => test);
  }

  if (!testCases.length) {
    console.log('No tests found!');
    process.exit();
  }

  testCases.forEach(testCase => {
    const result = testCase.run();

    if (outputType === 'RESULT') {
      console.log(result);
    } else if (outputType === 'TEST') {
      if (result === testCase.expected) {
        console.log(`[V] Passed test ${testCase.id}`);
      } else {
        console.log(`[X] Failed test ${testCase.id}`);
        console.log(`Expected: ${testCase.expected}`);
        console.log(`Got: ${result}`);
      }
    }
  });

  process.exit();
}

if (process && process.argv && process.argv.includes('-t')) {
  const onlyOutput = process.argv.includes('-o');
  const silent = process.argv.includes('-s');
  const indexOfT = process.argv.indexOf('-t');
  const testToRun = process.argv[indexOfT + 1];
  let outputType = 'TEST';

  if (onlyOutput) {
    outputType = 'RESULT';
  }

  if (silent) {
    outputType = 'SILENT';
  }

  VERBOSE = process.argv.includes('-v');

  return test(outputType, testToRun);
}

readLines();

module.exports = toposort;
